### PR TITLE
Parameterize the common-tag-input label and implement values upload / download

### LINKED ISSUE #1811: FF: Common-tag-input label is hard-coded as "tags", should be paramaterized.
Common-tag-input label is hard-coded as "tags", should be paramaterized. This came up in the add-list modal, the label should be `values`, but there's no way to change it from `tags`.

Note: reader beware, I tried to do this and it's either not as straightforward as it looks or I was overlooking something easy. hopefully the latter.

Image

### LINKED ISSUE #1812: FF fix: Enable the upload / download button functionality in Upsert-List values
Enable the upload / download button functionality in Upsert-List values. The Upsert-List modal utilizes the common-tag-input component and we should need to turn it on and test that it's operating correctly.

### PR DESCRIPTION
**Pull request by zackcl**
Resolves #1811, resolves #1812

The input label and placeholder are now dynamically set based on the input type ("tags" or "values").

The values upload / download functionality is implemented.

### CODE REVIEW

**Comment by zackcl** [comment]
I can also modify the code to make the label and placeholder be statically read from the `en.json` file if that's the convention we use. These will still need to be set conditionally in HTML I guess.

**Comment by zackcl** [comment]
I applied the changes I mentioned and also implemented the upload / download button functionality.

**Comment by zackcl** [comment]
This PR is now ready to be reviewed. Feel free to approve it.

**Comment by zackcl** [comment]
This PR is now ready. Please feel free to review and approve it.

**Comment by zackcl** [comment]
This PR is now ready. Please feel free to review and approve it.

**Comment by zackcl** [comment]
@danoswaltCL Could you please review this PR? Thank you!

File: frontend/projects/upgrade/src/app/features/dashboard/segments/modals/upsert-private-segment-list-modal/upsert-private-segment-list-modal.component.html [hunk 1]
```diff
@@ -22,21 +22,41 @@
...
+        [forceValidation]="forceValidation"
+        (downloadRequested)="onDownloadRequested($event)"
+      ></app-common-tags-input>
       <mat-form-field appearance="outline">
         <mat-label class="ft-14-400">Name</mat-label>
-        <input matInput formControlName="name" class="ft-14-400" />
+        <input matInput formControlName="name" placeholder="e.g., My schools" class="ft-14-400" />
+        <mat-hint class="form-hint ft-12-400">
+          {{
+            (config.title === 'Add Include List'
```
Comment (danoswaltCL) [inline frontend/projects/upgrade/src/app/features/dashboard/segments/modals/upsert-private-segment-list-modal/upsert-private-segment-list-modal.component.html:54]: translation is definitely something that I am not an expert on good practices for, but I think what we should be doing probably is passing these keys directly from the common-config, so extend that config type to include a 'hint' property (or whatever you want to name it), send in `hint: 'feature-flags.upsert-include-list-modal.name-hint.text'` and then we can do `{{ config.hint | translate }}` without needing this template logic.

File: frontend/projects/upgrade/src/app/shared-standalone-component-lib/components/common-tag-input/common-tag-input.component.html [hunk 2]
```diff
@@ -21,16 +33,37 @@
         [matChipInputSeparatorKeyCodes]="separatorKeysCodes"
         [matChipInputAddOnBlur]="addChipOnBlur"
         (matChipInputTokenEnd)="addChip($event)"
+        (focus)="onFocus()"
+        (blur)="onBlur()"
+        [placeholder]="tagsExist ? '' :
+          (inputType === CommonTagInputType.TAGS
```
Comment (danoswaltCL) [inline frontend/projects/upgrade/src/app/shared-standalone-component-lib/components/common-tag-input/common-tag-input.component.html:39]: same as other note about passing these in from config

**Comment by danoswaltCL** [comment]
I scanned this and put in some quick comments but saw much more complicated code than I was expecting in handling the file inputs, my sense is that it got complicated trying to be 'observable' friendly? but I'd have to sit down with this to understand it, I have a goal this week of getting something done with the "mooclet" stuff and reviews are going to be hard for me to deep dive into, if someone else can help.

**Comment by danoswaltCL** [comment]
"Maybe it's better to get rid of @Input() actionButtons = false; and have this logic be handled by @Input() inputType:? I think this will make the code simpler. @danoswaltCL Please let me know what you think about this (when you have time)."

Maybe, I'm actually wondering if the 'import/export' button(s) should just be it's own component, or even one component each, if there's non-trivial logic involved. Logic to upload / download stuff is feeling a little out of place in an input component, which already has a fairly complex primary job to managing tag behaviors, and it's only for one `InputType` mode, VALUES will never utilize any of that functionality.

**Comment by zackcl** [comment]
Thank you! I applied your suggested changes. Please feel free to approve this PR.

> Maybe, I'm actually wondering if the 'import/export' button(s) should just be it's own component, or even one component each, if there's non-trivial logic involved.

Let's consider this later maybe.

**Comment by zackcl** [comment]
I realized I need to apply one more change. Give me one minute.

**Comment by zackcl** [comment]
@danoswaltCL Done, please feel free to review and approve this PR.

File: frontend/projects/upgrade/src/app/shared-standalone-component-lib/components/common-modal/common-modal.types.ts [hunk 3]
```diff
@@ -23,3 +23,10 @@ export interface SimpleConfirmationModalParams {
   subMessage?: string;
   subMessageClass?: 'info' | 'warn';
 }
+
+export enum ModalSize {
```
Comment (danoswaltCL) [inline frontend/projects/upgrade/src/app/shared-standalone-component-lib/components/common-modal/common-modal.types.ts:27]: cool

File: frontend/projects/upgrade/src/app/shared-standalone-component-lib/components/common-tag-input/common-tag-input.component.ts [hunk 4]
```diff
@@ -42,42 +47,145 @@ import { TranslateModule } from '@ngx-translate/core';
...
+      this.showImportHelper = true;
+    }
+  }
+
+  onCloseButtonClick(event: MouseEvent): void {
+    event.preventDefault();
+    this.showImportHelper = false;
+  }
+
+  handleFilesSelected(files: File[]) {
```
Comment (danoswaltCL) [inline frontend/projects/upgrade/src/app/shared-standalone-component-lib/components/common-tag-input/common-tag-input.component.ts:120]: I think the main thing hanging me up from approving is just how complicated it is to read this part of the code. It was helpful in the demo to understand what this code is actually doing, so I think it's just a readability thing. I just don't know if I've ever seen anyone in a project like this actually have to create an Observable from scratch, usually we don't _need_ to but maybe that's cool, as long as everyone can understand what's going with this.

I think what would help a lot be to just simply break out the anonymous callback functions into their own named functions. Aka like the `reduce` callback, the `next` callback, the `onload` functions can all be broken out and named and commented a little. The error callbacks could be too but those are pretty simple.

**Comment by zackcl** [comment]
I applied your suggested change. Please feel free to approve this PR.

**Comment by danoswaltCL** [comment]
@zackcl i'd approve but i think the conflicts prob need resolved first, then i can approve and merge this one

**Comment by zackcl** [comment]
@danoswaltCL I just found out that in the current dev branch, I get the following error when I try to delete an include list:
```
app-error-handler.service.ts:18 ERROR 
TypeError: Cannot read properties of undefined (reading 'some')
    at feature-flags.reducer.ts:177:54
    at Array.find (<anonymous>)
    at feature-flags.reducer.ts:176:48
    at ngrx-store.mjs:1565:26
    at featureFlagsReducer (feature-flags.reducer.ts:280:10)
    at ngrx-store.mjs:377:20
    at combination (ngrx-store.mjs:332:37)
    at clear-state.reducer.ts:27:12
    at init-state-from-loca…age.reducer.ts:8:22
    at debug.reducer.ts:7:22
```
But this error doesn't seem to happen in this PR's branch, which might mean that some of the recent changes merged into the dev branch are causing this error. I will investigate further before resolving conflicts and merging the dev branch into this one.

**Comment by zackcl** [comment]
It looks like the error happens sometimes, but not always. It seems to be working again now.

**Comment by danoswaltCL** [comment]
ok sure, i won't merge anything else into dev, let me know if i can assist

**Comment by zackcl** [comment]
@danoswaltCL The error I mentioned had nothing to do with this PR. Please fee free to approve this. Thanks!

**Review by danoswaltCL — APPROVED** [review]

**Pull request merged**